### PR TITLE
Adopt type hinting and tidy up style

### LINKED ISSUE #19: Fully adopt typing
Since we have migrated to Python 3 we should take advantage of its features. In particular this issue is interested in type hinting. This brings some of the riguour of statically typed languages to Python, much to the advantage of long term maintenance.

This change should also include the adoption of a type checking tool in our action pipeline.

### PR DESCRIPTION
**Pull request by MatthewHambley**
This change adopts type hinting and the use of MyPy to check it. It also tidies up a few matters of code style.

Closes #19

### CODE REVIEW

**Review by EdHone — COMMENTED** [review]
Everything here looks pretty good, I only have a few minor in-line comments that I've added. We briefly discussed adding more fail-state unit-tests for reading of style files but that is probably out of the scope for this pull request. 
Please have a look at the comments I've added and pass the ticket back to me.

File: source/stylist/__main__.py [hunk 1]
```diff
@@ -49,7 +50,17 @@ def parse_cli() -> argparse.Namespace:
...
     cli_parser.add_argument('-help', '-h', '--help', action='help')
     cli_parser.add_argument('-verbose', '-v', action="store_true",
-                            help='Produce a running commentary on progress')
+                            help="Produce a running commentary on progress.")
+    cli_parser.add_argument('-configuration',
+                            type=Path,
+                            default=None,
+                            metavar='FILENAME',
+                            help="File which configures the tool.")
+    help = "Style to use for check. May be specified repeatedly."
```
[the remark this answers is not in the document]
Comment (MatthewHambley) [reply source/stylist/__main__.py:59]: I've removed the short form for the verbose switch.

Help is special, it covers as many commonly used variants as I can think of in order that people can always get help.

File: source/stylist/source.py [hunk 2]
```diff
@@ -299,120 +305,142 @@ def path(self,
...
                    indent: int = 0) -> None:
-        '''
+        """
         Dumps a textual representation of the tree to standard out.
         Intended for debug use.
-        '''
+        """
+        # Argument "children" confuses Pycharm as it does not appear as
+        # iterable even though it is. MyPy is not bothered.
+        #
```
[the remark this answers is not in the document]
Comment (MatthewHambley) [reply source/stylist/source.py:386]: It's only a problem if you aren't expecting it. As far as I can tell the issue is that fparser does not implement type hinting. When that changes I would expect this issue to go away.

File: system-tests/simple/naughty.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+The first line of this file is okay.
+But this next line has trailing whitespace. 	 
+The fine line is okay though.
```
[the remark this answers is not in the document]
Comment (MatthewHambley) [reply system-tests/simple/naughty.txt]: Fixed.

**Review by EdHone — APPROVED** [review]
Thanks, Matthew I'm happy that this **passes** my code review.

**Pull request merged**